src/lib.rs

```diff
@@ -8,72 +8,120 @@
 //! use std::fmt::Write;
 //! let mut buffer: WriteBuffer<20> = WriteBuffer::new();
 //! let x = 12;
-//! write!(buffer, "{}", x).expect("Can't write");
-//! assert_eq!(buffer.as_str().unwrap(), "12");
+//! write!(buffer, "{}", x).unwrap();
+//! assert_eq!(buffer.as_str(), "12");
 //! ```
 
 use core::fmt::{self, Display, Formatter};
 
 /// A write buffer
 #[derive(Debug)]
 pub struct WriteBuffer<const N: usize> {
-    buf: [u8; N],
-    offset: usize,
+    buffer: [u8; N],
+    cursor: usize,
 }
 
 impl<const N: usize> WriteBuffer<N> {
     /// Creates a write buffer
     pub fn new() -> Self {
         let buf = [0u8; N];
-        WriteBuffer { buf, offset: 0 }
+        WriteBuffer {
+            buffer: buf,
+            cursor: 0,
+        }
     }
 
-    /// Returns the valid portion of the buffer as slice
+    /// Returns a slice containing the already written bytes in the buffer
     pub fn as_slice(&self) -> &[u8] {
-        &self.buf[..self.offset]
+        &self.buffer[..self.cursor]
+    }
+
+    /// Returns a mutable slice containing the already written bytes in the
+    /// buffer
+    ///
+    /// Dev Note: This should _not_ be `pub`, since otherwise the user might
+    /// mess with the bytes, violating the guarantee that the safety of
+    /// [`as_str`] and [`as_str_mut`] depend on!
+    fn as_slice_mut(&mut self) -> &mut [u8] {
+        &mut self.buffer[..self.cursor]
     }
 
-    /// Reset the buffer to be reused
+    /// Reset the buffer such that it can be reused.
+    ///
+    /// Note: This does _not_ overwrite any data in memory, it only sets the
+    /// internal cursor back to the start of the buffer.
     pub fn reset(&mut self) {
-        self.offset = 0;
+        self.cursor = 0;
     }
 
-    /// Converts the buffer in `&str`, returnig None if some bytes doesn't resemble valid utf-8
-    pub fn as_str(&self) -> Option<&str> {
-        core::str::from_utf8(self.as_slice()).ok()
+    /// Converts the buffer into `&str`.
+    pub fn as_str(&self) -> &str {
+        // SAFETY: The only way to write into `self.buf` is via
+        // `Write::write_str`. Therefore it is always guaranteed that the buffer
+        // contains valid UTF-8.
```

Comment:
The method `as_slice_mut` allows to change the buffer to invalid utf-8

Reply:
That's why it's not `pub`, it only exists to make the internal API cleaner :)